### PR TITLE
[refactor] moe: compute softmax after argmax

### PR DESCRIPTION
**Pull request by msbaines**
According to @TimDettmers:

So the flow from the balancing loss to the gate is the same. The
difference is when we multiply the softmax probability by the
expert we have a softmax gradient flow from the expert to the gate.
The problem with that is that (1) softmax gradients are small since
the gradient is p*(1-p), (2) if the gradient is close to 0 then the
deltas/errors coming from the expert are multiplied by this value on
the way to the gate. This means that a probability close to zero
should be avoided or otherwise errors from the 2nd expert will not
update the gate and the gate "does not know about the improvement to
the expert". Adam will help with this expect a little bit, since the
running differences for such an expert are small but we only get on
average every k/num_experts samples an running differences update
for adam. So adam might not be able to adjust quickly enough for a
situation like this.

The second piece is since we normalize in the softmax by the sum of
all exponentiated terms, and each term is positive, the normalizing
terms grows with the size of the number of experts. So if we have
1024 experts, it would be very likely that most of the probability
terms are close to zero and hence have very small gradients. If we
just take the top-k values and normalize them, then the normalizing
term is much smaller and the minimum probability is still sizable to
allow gradient flow from expert to gate.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by min-xu-ai — COMMENTED** [review]
maybe I am missing something. the new version of the code doesn't need a softmax anymore? Or is it done elsewhere?

**Comment by msbaines** [comment]
> maybe I am missing something. the new version of the code doesn't need a softmax anymore? Or is it done elsewhere?

Softmax is done "by hand", see line 67-69.

File: fairscale/nn/moe/top2gate.py [hunk 1]
```diff
@@ -71,8 +64,8 @@ def top2gating(logits: torch.Tensor) -> Tuple[Tensor, Tensor, Tensor]:
     # Normalize gate probabilities
     mask1_float = mask1.float()
     mask2_float = mask2.float()
-    gates1_gs = torch.einsum("gse,gse->gs", gates, mask1_float)
-    gates2_gs = torch.einsum("gse,gse->gs", gates, mask2_float)
+    gates1_gs = torch.einsum("gse,gse->gs", torch.exp(logits), mask1_float)
+    gates2_gs = torch.einsum("gse,gse->gs", torch.exp(logits), mask2_float)
```
Comment (min-xu-ai) [inline fairscale/nn/moe/top2gate.py:68]: will torch be smart enough and perform a single torch.exp on logits here or it will be done twice on the same data?
